### DISCUSSION TITLE
How do I decode bytearray for power data

### DISCUSSION THREAD

**Discussion by EkoFreshFTech** [hbldh/bleak#1579] [Q&A] [answered]
* bleak version: 0.22.1
* Python version: 3.11.9
* Operating System: Windows

### Description
Hi,
I wrote a script to recieve data via async callback. My problem is that I get unusual ouputs. I tried to print the power from my hometrainer.

```python
from bleak import BleakClient
import asyncio

async def callback_data(sender, data):

    print("Rohdaten:", data)

async def connect_and_data():

    cyc_power_uuid = '00002a63-0000-1000-8000-00805f9b34fb'
    address = 'C5:B3:F0:FB:9A:1E'
    try:
        async with BleakClient(address) as client:
            await client.start_notify(cyc_power_uuid, callback_data)
            await asyncio.sleep(10)
            await client.stop_notify(cyc_power_uuid)
    except Exception as e:
        print(f"Fehler beim Verbinden oder Lesen: {e}")
    except KeyboardInterrupt:
        await client.stop_notify(cyc_power_uuid)

# Starte die asynchrone Funktion
asyncio.run(connect_and_data())
```
Output:
```
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
Rohdaten: bytearray(b'1\x00\x00\x004\xeb\x02\x00\x00\x00p \x00r\x1b')
```
How do I decode this?
Thanks for any help :D

PS: is there any chance of increasing the rate of data sent? Right now i get 1 Hz, but i try to reach around 5 Hz.

**Comment by koenvervloesem** [comment, marked as answer]
You can find the representation of the data in the [GATT Specification Supplement](https://www.bluetooth.com/specifications/gss/). Search for "Cycling Power Measurement" in the document.

**Reply by EkoFreshFTech** [reply]
I read the documentation but i dont know how to identify the 2 Bytes long Flag. I know the flag determines what type of data is in the bytearray, but how do I know what bit is 0 or 1 in the flag boolean[16]?
F.e. `bytearray(b'1\x01\x00\x00z\x9f\x00\x00\x00D!\x07\x00\xe1x')` where is here the flag and what information do i get out of it?

**Reply by koenvervloesem** [reply]
The bits of the Flags field are described on page 81.

**Reply by dlech** [reply]
I would use an enum to make it easy to read.

```python
from enum import IntFlag

class CyclingPowerFlags(IntFlag):
    PEDAL_POWER_BALANCE_PRESENT = 1<<0
    PEDAL_POWER_BALANCE_REFERENCE = 1<<1
    ACCUMULATED_TORQUE_PRESENT = 1<<2
    ACCUMULATED_TORQUE_SOURCE = 1<<3
    WHEEL_REVOLUTION_DATA_PRESENT = 1<<4
    CRANK_REVOLUTION_DATA_PRESENT =1<<5
    EXTREME_FORCE_MAGNITUDES_PRESENT = 1<<6
    EXTREME_TORQUE_MAGNITUDES_PRESENT = 1<<7
    EXTREME_ANGLES_PRESENT = 1<<8
    TOP_DEAD_SPOT_ANGLE_PRESENT = 1<<9
    BOTTOM_DEAD_SPOT_ANGLE_PRESENT = 1<<10
    ACCUMULATED_ENERGY_PRESENT = 1<<11
    OFFSET_COMPENSATION_INDICATOR = 1<<12

def callback_data(sender, data: bytes):
    flags = CyclingPowerFlags(int.from_bytes(data[:2], byteorder="little"))
    print("Flags:", flags)
```

Then you can write something to read the remaining info based on the flags. I would probably wrap the data in a `BytesIO` object so I can just `read()` from it instead of keeping track of a position.

**Comment by dlech** [comment]
> : is there any chance of increasing the rate of data sent?

It is up to the peripheral device. There isn't a standard way to do this that I know of. You could use the service explorer sample program to see if there are any custom characteristics that seem like they might allow changing the rate. But if not, there might not be a way to change it.

**Reply by EkoFreshFTech** [reply]
I got plenty of different characteristics regarding Cycling Power, but none of those can be used for the notification method (except Handle 23, which is sending with 1 Hz).
```
[Service] %s 00001818-0000-1000-8000-00805f9b34fb (Handle: 22): Cycling Power
  [Characteristic] %s (%s)%s 00002a63-0000-1000-8000-00805f9b34fb (Handle: 23): Cycling Power Measurement notify
    [Descriptor] %s, Value: %r 00002902-0000-1000-8000-00805f9b34fb (Handle: 25): Client Characteristic Configuration bytearray(b'\x00\x00')
  [Characteristic] %s (%s)%s 00002a65-0000-1000-8000-00805f9b34fb (Handle: 26): Cycling Power Feature read , Value: bytearray(b'\r\x02\x00\x00')
  [Characteristic] %s (%s)%s 00002a5d-0000-1000-8000-00805f9b34fb (Handle: 28): Sensor Location read , Error: Multiple Characteristics with this UUID, refer to your desired characteristic by the `handle` attribute instead.
  [Characteristic] %s (%s)%s 00002a66-0000-1000-8000-00805f9b34fb (Handle: 30): Cycling Power Control Point write,indicate
    [Descriptor] %s, Value: %r 00002902-0000-1000-8000-00805f9b34fb (Handle: 32): Client Characteristic Configuration bytearray(b'\x00\x00')
```

**Reply by EkoFreshFTech** [reply]
Hi,
you were right! The hometrainer I'm using has a feature called 'Fast Track', that allows a data rate of 10 Hz. Just had to install the app for the trainer on my phone, activate it and now with bleak I can send data with a rate of 10 Hz (the uuid is still Cycling Power Measurement notify, see above)
Thanks for your help :D
